tbcontrol/symbolic.py

```diff
@@ -53,6 +53,15 @@ def routh(p):
         for j in range(N//2):
             S = M[[i-2, i-1], [0, j+1]]
             M[i, j] = sympy.simplify(-S.det()/M[i-1,0])
+        # If a row of the routh table becomes zero,Take the derivative of the previous row and substitute it instead
+        # Ref: Norman S. Nise, Control Systems Engineering, 8th Edition, Chapter 6, Section 3
+        if M[i,:] == sympy.Matrix([[0]*(M.shape[1])]):
+            #Find the coefficients on taking the derivative
+            diff_arr = numpy.arange(N-i, -0.1, -2, dtype=int)
```

Comment:
I don't understand how this is taking the derivative as mentioned in the comments.

Reply:
```python
if M[i,:] == sympy.Matrix([[0]*(M.shape[1])]):
```
Checks if the entire row of the routh array is zero
```python
diff_arr = numpy.arange(N-i, -0.1, -2, dtype=int)
```
We need to differentiate the auxiliary eqn $A(s)$. The order of $A(s)$ is $N-1$ .The  auxiliary eqn is of the form $k_1s^{N-1} + k_2s^{N-3} + k_3s^{N-5} ...$ hence differentiation of this is $(N-1)  k_1s^{N-2} + (N-3)k_2s^{N-4} + (N-5)k_3s^{N-6} ...$  Ref [this](https://en.wikipedia.org/wiki/Routh%E2%80%93Hurwitz_stability_criterion#Routh%E2%80%93Hurwitz_criterion_for_second,_third_and_fourth-order_polynomials:~:text=is%20called%20the%20%22-,auxiliary%20polynomial,-%22.). The above code creates a NumPy array with terms `[N-1,N-3,N-5 ...]` 
Dot Multiplying this with the Coefficients of $A(s)$ is differentiation. This works because the Coefficients of the polynomial $A(s)$ is never a function of $s$